### PR TITLE
[Snyk] Security upgrade keras from 2.11.0 to 3.12.0

### PR DESCRIPTION
**Pull request by jhamot**
snyk-top-banner

#### Snyk changed the following file(s):

- `tf2/requirements.txt`

<details>
<summary>⚠️ <b>Warning</b></summary>

```
tensorflow 2.11.0 requires keras, which is not installed.
tensorflow 2.11.0 requires tensorflow-io-gcs-filesystem, which is not installed.
tensorflow-metadata 1.12.0 has requirement absl-py<2.0.0,>=0.9, but you have absl-py 2.1.0.
tensorboard 2.11.2 has requirement setuptools>=41.0.0, but you have setuptools 40.5.0.
googleapis-common-protos 1.72.0 has requirement protobuf!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<7.0.0,>=3.20.2, but you have protobuf 3.19.6.
```

</details>

#### Breaking Change Risk

Merge Risk: High

> **Notice:** This assessment is enhanced by AI.

---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.
> - Some vulnerabilities couldn't be fully fixed and so Snyk will still find them when the project is tested again. This may be because the vulnerability existed within more than one direct dependency, but not all of the affected dependencies could be upgraded.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: 
🧐 [View latest project report](https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Directory Traversal](https://learn.snyk.io/lesson/directory-traversal/?loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"breakingChangeRiskLevel":"high","FF_showPullRequestBreakingChanges":false,"FF_showPullRequestBreakingChangesWebSearch":false,"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"keras","from":"2.11.0","to":"3.12.0"}],"env":"prod","issuesToFix":["SNYK-PYTHON-KERAS-14152002"],"prId":"ec8d15a1-1ee9-43a5-b864-f098e41a07fa","prPublicId":"ec8d15a1-1ee9-43a5-b864-f098e41a07fa","packageManager":"pip","priorityScoreList":[142],"projectPublicId":"c1bd5c26-cdb9-4fef-9b86-a38cffefa71b","projectUrl":"https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","pr-warning-shown","priorityScore"],"type":"auto","upgrade":[],"vulns":["SNYK-PYTHON-KERAS-14152002"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

### CODE REVIEW

**Comment by jhamot** [comment]
Merge Risk: High

The upgrade from Keras 2.11.0 to 3.12.0 is a major architectural overhaul, introducing multi-backend support for TensorFlow, JAX, and PyTorch. This is a significant change that requires code modifications and careful validation.  
  
**Key Breaking Changes:**  
  
*   **Multi-Backend Architecture**: Keras 3 is a full rewrite designed to be backend-agnostic. While it aims for high backward compatibility, custom components or direct use of backend-specific APIs will require refactoring.  
*   **Package Imports**: All imports must be changed. Code that used `from tensorflow.keras import ...` must be updated to `from keras import ...`.  
*   **Backend-Agnostic Operations**: To write code that works across all backends, you must replace TensorFlow-specific functions (e.g., `tf.matmul`) with their backend-agnostic equivalents in the `keras.ops` namespace (e.g., `keras.ops.matmul`).  
*   **Model Saving**: Saving models in the TensorFlow `SavedModel` format via `model.save()` is no longer supported. The recommended format is `.keras` v3.  
*   **JIT Compilation by Default**: The `jit_compile` argument in `Model` is now `True` by default on GPUs. This can improve performance but may cause errors with custom layers or operations not supported by XLA.  
*   **Custom Variable Tracking**: In custom layers, `tf.Variable` attributes are no longer automatically tracked. You must use `self.add_weight()` or `keras.Variable` to ensure variables are registered correctly.  
  
**Recommendation:**  
This is a high-risk upgrade that requires significant code changes. Before merging, developers must:  
1.  Update all `tensorflow.keras` imports to `keras`.  
2.  Review and refactor any custom layers or models that use TensorFlow-specific APIs to use the `keras.ops` namespace.  
3.  Verify model saving and loading logic to ensure compatibility with the new `.keras` format.  
4.  Test thoroughly, paying close attention to potential XLA-related errors from the new default JIT compilation.  
  
Source: [Keras 2 to 3 Migration Guide](https://keras.io/guides/migrating_to_keras_3/), [Keras 3.0 Announcement](https://keras.io/keras_3/)

> **Notice 🤖:** This content was augmented using **artificial intelligence**. AI-generated content **may** contain **errors** and should be **reviewed** for accuracy before use.
